### PR TITLE
Add dockerfile

### PR DESCRIPTION
**Pull request by AnushreeBannadabhavi**
Initial Version of the DockerFile. Build in-progress. This PR contains all the changes from PR #14 and PR #13

### CODE REVIEW

**Comment by uditagarwal97** [comment]
I'm getting the following error in the dockerfile:
``` ---> cf1867d76772
Step 11/21 : RUN unzip protobuf-all-3.17.2.zip
 ---> Running in 04d0d1e1620f
/bin/sh: 1: unzip: not found
The command '/bin/sh -c unzip protobuf-all-3.17.2.zip' returned a non-zero code: 127
```

**Comment by AnushreeBannadabhavi** [comment]
I'll add "RUN apt-get install unzip" in the Dockerfile. That should hopefully fix the issue. I'll also add all necessary changes in the next commit after verifying the created image.

File: docker/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+RUN unzip protobuf-all-3.17.2.zip
+WORKDIR /home/LLTFI/protobuf-3.17.2
+
+RUN ./configure
+RUN make
+RUN make check
+RUN make install
+RUN ldconfig
+
+WORKDIR /home/LLTFI/
```
Comment (uditagarwal97) [inline docker/Dockerfile:82]: WORKDIR /home/

File: docker/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+    cd .. && \
+    export MLIR_DIR=$(pwd)/llvm-project/build/lib/cmake/mlir && \
+    mkdir onnx-mlir/build && cd onnx-mlir/build && \
+    if [[ -z "$pythonLocation" ]]; then \
+  	cmake -G Ninja \
+        -DCMAKE_CXX_COMPILER=/usr/bin/c++ \
+        -DMLIR_DIR=${MLIR_DIR} \
+        .. \
+    else \
+  	cmake -G Ninja \
```
Comment (uditagarwal97) [inline docker/Dockerfile:97]: Remove the else part of this if-else condition. Command to build onnx-mlir:
``` 
cmake -G Ninja 
        -DCMAKE_CXX_COMPILER=/usr/bin/c++ 
        -DMLIR_DIR=${MLIR_DIR} 
        .. 
```

File: docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        -DCMAKE_CXX_COMPILER=/usr/bin/c++ \
+        -DPython3_ROOT_DIR=$pythonLocation \
+        -DMLIR_DIR=${MLIR_DIR} \
+        .. \
+    fi && \
+    cmake --build . -j 2 && \
```
Comment (uditagarwal97) [inline docker/Dockerfile:103]: cmake --build . -j ${nproc} && \

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    export LIT_OPTS=-v && \
```
Comment (uditagarwal97) [inline docker/Dockerfile:104]: No need for this command 'LIT_OPS'

File: docker/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    cmake --build . --target check-onnx-lit && \
```
Comment (uditagarwal97) [inline docker/Dockerfile:105]: remove 'this command as well. Ninja install will be sufficient

File: docker/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    ninja install
+
+WORKDIR /home/LLTFI/
+
+### LLTFI
+RUN git clone https://github.com/DependableSystemsLab/LLTFI.git && \
+    cd /home/LLTFI/LLTFI && \
```
Comment (uditagarwal97) [inline docker/Dockerfile:112]: Add command 'pip install pyyaml'

File: docker/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    ./setup -LLFI_BUILD_ROOT $(pwd)/build -LLVM_SRC_ROOT $(pwd)/../llvm-project -LLVM_DST_ROOT $(pwd)/../llvm-project/build --no_gui && \
```
Comment (uditagarwal97) [inline docker/Dockerfile:113]: LLVM_SRC_ROOt is wrong. it should be '$(pwd)/../../llvm_project'. same with LLVM_DST_ROOT

File: docker/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+          # Use same versions as those in ubuntu:focal
+          pip3 install -q \
+               Cython pytest==4.6.9 pytest-forked==1.1.3 \
+               pytest-xdist==1.31.0 typing-extensions==3.7.4.1 && \
+          rm -rf /var/cache/dnf/* && \
+          # Install ninja
+          git clone -b v1.10.2 https://github.com/ninja-build/ninja.git && \
+          cd ninja && mkdir -p build && cd build && \
+          cmake .. && \
+          make -j${NPROC} install \
```
Comment (uditagarwal97) [inline docker/Dockerfile:48]: Add a semicolon. ' make -j${NPROC} install; '

File: docker/README.md [hunk 9]
```diff
@@ -0,0 +1,9 @@
+### Building LLTFI using Docker: ###
+
+`docker/Dockerfile` can be used to build and run LLTFI in a docker container.
+- Use the variable `NPROC` to specify the number of threads that should run simultaneously. 
+- All dependencies required to build LLTFI and run sample programs and tests will be installed in the docker container built using this docker file.
```
Comment (uditagarwal97) [inline docker/README.md:5]: Mention that the docker image does not contain dependencies for the GUI.

File: README.md [hunk 10]
```diff
@@ -1,12 +1,32 @@
...
+As machine learning (ML) has become more prevalent across many critical domains, so has the need to understand ML system resilience. While there are many ML fault injectors at the application level, there has been little work enabling fault injection of ML applications at a lower level. **LLTFI** is a tool that allows users to run fault injection experiments on C/C++, TensorFlow and PyTorch applications at a lower level (at the LLVM IR level). Please refer to the following [paper](https://blogs.ubc.ca/dependablesystemslab/2021/08/31/wip-lltfi-low-level-tensor-fault-injector/) for more information about LLTFI.
 
-Because LLTFI is designed to be backwards compatible with LLFI, the basic setup instructions for LLTFI are the same as those of LLFI.
-**For instructions on how to setup and run LLTFI on ML programs, please read [this page](sample_programs/mnist).**
+LLTFI is built on top of [LLFI](https://github.com/DependableSystemsLab/LLFI) and is fully backwards compatible with LLFI.
```
Comment (uditagarwal97) [inline README.md:7]: LLTFI is built on top of [LLFI](https://github.com/DependableSystemsLab/LLFI) and is fully backwards compatible with it.

File: README.md [hunk 11]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
-  9. GraphViz package (for visualizing error propagation)
+  1. 64 Bit Machine (preferably with GPU for faster training of ML programs) 
+  2. 64 bit Linux (Ubuntu 20.04) or OS X
+  3. CMake (minimum v3.15)
+  4. Python 3 and above
+  5. Python YAML library (PyYAML)
+  6. Ninja >= 1.10.2
+  7. libprotoc >= 3.11.0
+  8. Clang v13.0 (commit: 23dd750279c9)
+  9. LLVM v13.0 (commit: 23dd750279c9) ([Reference](http://llvm.org/docs/CMake.html)).
```
Comment (uditagarwal97) [inline README.md:77]: Update the Clang and LLVM version in the README file.

File: README.md [hunk 12]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
+
+		    The MLIR_DIR cmake variable must be set before building onnx-mlir. It should point to the mlir cmake module inside an llvm-project build or install directory (e.g., llvm-project/build/lib/cmake/mlir).
+            ```
+		    MLIR_DIR=$(pwd)/llvm-project/build/lib/cmake/mlir
+		    ```
+
+		    Onnx-mlir commit: ``` 221b8e1d2ad ``` has to be built and installed. 
+			```
+			git clone --recursive https://github.com/onnx/onnx-mlir.git
+			cd onnx-mlir && git checkout 221b8e1d2ad && cd ..
```
Comment (uditagarwal97) [inline README.md:129]: Use the lastest version of ONNX-MLIR

File: README.md [hunk 13]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
+  6. Ninja >= 1.10.2
+  7. libprotoc >= 3.11.0
+  8. Clang v13.0 (commit: 23dd750279c9)
+  9. LLVM v13.0 (commit: 23dd750279c9) ([Reference](http://llvm.org/docs/CMake.html)).
+		LLVM 13.0 takes a long time to completely build. Following is a shortcut to checking out the required LLVM commit, and building only the necessary LLVM targets.
+		```
+		git clone https://github.com/llvm/llvm-project.git
+		
+		# Check out a specific branch that is known to work with ONNX MLIR.
+		cd llvm-project && git checkout 23dd750279c9 && cd ..
```
Comment (uditagarwal97) [inline README.md:83]: Use the latest version of LLVM

File: README.md [hunk 14]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
+		cd llvm-project/build
+		
+		cmake -G Ninja ../llvm \
+			-DLLVM_ENABLE_PROJECTS="clang;mlir;tools" \
+			-DLLVM_BUILD_TESTS=ON \
+			-DLLVM_TARGETS_TO_BUILD="host" \
+			-DLLVM_ENABLE_ASSERTIONS=ON \
+			-DLLVM_ENABLE_RTTI=ON
+
+		cmake --build . --target clang check-mlir mlir-translate opt llc lli llvm-dis llvm-link
```
Comment (uditagarwal97) [inline README.md:95]: Add `-j2` flag here.

File: README.md [hunk 15]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
+		    pip install tf2onnx
+		    ```
+		4. libprotoc
+			```
+			curl -OL https://github.com/protocolbuffers/protobuf/releases/download/v3.17.2/protobuf-all-3.17.2.zip
+			unzip protobuf-all-3.17.2.zip
+			cd protobuf-3.17.2
+			
+			./configure
+			make
```
Comment (uditagarwal97) [inline README.md:114]: Add `-j2` flag here as well

File: README.md [hunk 16]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
+		    ```
+
+		    Onnx-mlir commit: ``` 221b8e1d2ad ``` has to be built and installed. 
+			```
+			git clone --recursive https://github.com/onnx/onnx-mlir.git
+			cd onnx-mlir && git checkout 221b8e1d2ad && cd ..
+	
+			mkdir onnx-mlir/build && cd onnx-mlir/build
+			if [[ -z "$pythonLocation" ]]; then
+			cmake -G Ninja \
```
Comment (uditagarwal97) [inline README.md:133]: Remove the if-else condition. We only need the if part. Just like we have done in the Dockerfile.

File: README.md [hunk 17]
```diff
@@ -39,29 +59,106 @@ About tcsh:
...
+					-DCMAKE_CXX_COMPILER=/usr/bin/c++ \
+					-DPython3_ROOT_DIR=$pythonLocation \
+					-DMLIR_DIR=${MLIR_DIR} \
+					..
+			fi
+			cmake --build .
+				
+			# Run lit tests:
+			export LIT_OPTS=-v
```
Comment (uditagarwal97) [inline README.md:147]: Should we build the lit tests of ONNX-MLIR? If it takes too l;ong to build and run the test case, perhaps we can skip this step.

**Review by uditagarwal97 — CHANGES_REQUESTED** [review]

**Comment by AnushreeBannadabhavi** [comment]
Addressed all the review comments in commit id '7b06e8d50ecdc4bb32b5da053fbf35a1933dbbd5'

**Comment by AnushreeBannadabhavi** [comment]
Closed PR accidently

**Review by uditagarwal97 — APPROVED** [review]

**Review by abrahamchan — APPROVED** [review]
LGTM! I will go ahead and merge this as Udit also approved this too.

**Pull request merged**